fish-rust/src/fd_monitor.rs

```diff
@@ -0,0 +1,573 @@
+    //     /// Resets to a new fd, taking ownership.
+    //     fn reset(self: Pin<&mut AutoCloseFd>, fd: i32);
+    //     /// Returns `true` if this `AutoCloseFd` instance has a valid fd (`>= 0`).
+    //     #[rust_name = "is_valid"]
+    //     fn valid(&self) -> bool;
+    // }
+    extern "Rust" {
+        #[cxx_name = "fd_monitor_item_id_t"]
+        type FdMonitorItemId;
+    }
+
+    extern "Rust" {
+        #[cxx_name = "fd_monitor_item_t"]
+        type FdMonitorItem;
+
+        #[cxx_name = "make_fd_monitor_item_t"]
+        fn new_fd_monitor_item_ffi(fd: i32, timeout_usecs: u64, callback: *const c_void, param: *const c_void) -> Box<FdMonitorItem>;
+    }
+
+    extern "Rust" {
+        #[cxx_name = "fd_monitor_t"]
+        type FdMonitor;
+
+        #[cxx_name = "make_fd_monitor_t"]
+        fn new_fd_monitor_ffi() -> Box<FdMonitor>;
+
+        #[cxx_name = "add_item"]
+        fn add_item_ffi(&mut self, fd: i32, timeout_usecs: u64, callback: *const c_void, param: *const c_void) -> u64;
+
+        #[cxx_name = "poke_item"]
+        fn poke_item_ffi(&self, item_id: u64);
+
+        #[cxx_name = "add"]
+        pub fn add_ffi(&mut self, item: Box<FdMonitorItem>) -> u64;
+    }
+}
+
+// TODO: Remove once we're no longer using the FFI variants of FdEventSignaller and AutoCloseFd.
+unsafe impl Sync for FdEventSignaller {}
+unsafe impl Send for FdEventSignaller {}
+// unsafe impl Send for AutoCloseFd {}
+// unsafe impl Sync for AutoCloseFd {}
+
+// TODO: Remove once we're no longer using the FFI variant of AutoCloseFd.
+impl Unpin for FdMonitorItem {}
+
+/// Each item added to fd_monitor_t is assigned a unique ID, which is not recycled. Items may have
+/// their callback triggered immediately by passing the ID. Zero is a sentinel.
+#[repr(transparent)]
+#[derive(Clone, Copy, Debug, PartialEq, Eq, PartialOrd, Ord)]
+pub struct FdMonitorItemId(u64);
+
+type FfiCallback = extern "C" fn(*mut AutoCloseFd2, u8, void_ptr);
+
+/// The callback type used by [`FdMonitorItem`]. It is passed a mutable reference to the
+/// `FdMonitorItem`'s [`FdMonitorItem::fd`] and [the reason](ItemWakeupReason) for the wakeup. The
+/// callback may close the fd, in which case the `FdMonitorItem` is removed from [`FdMonitor`]'s
+/// set.
+///
+/// We can't use generics here and need to resort to dynamic dispatch in order to store multiple
+/// `FdMonitorItem` objects in the same vector (in [`FdMonitor`]), which isn't possible since
+/// each closure/callback is given its own type.
+///
+/// Additionally, as capturing C++ closures can't be safely used via ffi interop, we have a separate
+/// variant of the type that uses the C abi to invoke a callback. This will be removed when the
+/// dependent C++ code (currently only `src/io.cpp`) is ported to rust
+enum FdMonitorCallback {
+    None,
+    Native(Box<dyn Fn(&mut AutoCloseFd2, ItemWakeReason) + Send + Sync>),
+    Ffi(FfiCallback /* fn ptr */, void_ptr /* param */),
```

Comment:
The enum seems like a nice solution for isolating the FFI ugliness.

The callback context argument is also a good choice - simple and effective.